packages/core/src/fhir-to-cda/cda-templates/components/observations.ts

```diff
@@ -1,37 +1,251 @@
+import {
+  CdaCodeCv,
+  CdaInstanceIdentifier,
+  CdaValueSt,
+  Entry,
+  EntryObject,
+  ObservationEntry,
+  ObservationEntryRelationship,
+  ObservationTableRow,
+} from "../../cda-types/shared-types";
 import {
   TIMESTAMP_CLEANUP_REGEX,
+  buildCodeCe,
   buildCodeCvFromCodeableConcept,
   buildInstanceIdentifier,
-  buildInstanceIdentifiersFromIdentifier,
-  buildValueST,
+  buildReferenceId,
+  buildValueSt,
+  formatDateToCdaTimestamp,
+  formatDateToHumanReadableFormat,
+  isLoinc,
   withoutNullFlavorObject,
 } from "../commons";
-import { _codeAttribute, _valueAttribute } from "../constants";
+import {
+  _classCodeAttribute,
+  _codeAttribute,
+  _extensionAttribute,
+  _idAttribute,
+  _inlineTextAttribute,
+  _moodCodeAttribute,
+  _rootAttribute,
+  _typeCodeAttribute,
+  _valueAttribute,
+  extensionValue2015,
+  loincCodeSystem,
+  loincSystemName,
+  placeholderOrgOid,
+} from "../constants";
+import { AugmentedObservation } from "./augmented-resources";
+
+export interface CDAObservation {
+  component: {
+    observation: {
+      [_classCodeAttribute]: Entry;
+      [_moodCodeAttribute]: Entry;
+      id?: CdaInstanceIdentifier[] | Entry;
+      code: CdaCodeCv | Entry;
+      text?: Entry;
+      statusCode?: EntryObject;
+      effectiveTime?: {
+        low?: EntryObject;
+        high?: EntryObject;
+      };
+      priorityCode?: Entry;
+      // TODO support other types of values like CodeableConcept, Quantity, etc.
+      value?: CdaValueSt | undefined;
+    };
+  };
+}
 
 export function buildObservations(observations: Observation[]): CDAObservation[] {
   return observations.map(observation => {
     const effectiveTime = observation.effectiveDateTime?.replace(TIMESTAMP_CLEANUP_REGEX, "");
     return {
       component: {
         observation: {
-          _classCodeAttribute: "OBS",
-          _moodCodeAttribute: "EVN",
+          [_classCodeAttribute]: "OBS",
+          [_moodCodeAttribute]: "EVN",
           templateId: buildInstanceIdentifier({
             root: "2.16.840.1.113883.10.20.22.4.2",
-            extension: "2015-08-01",
+            extension: extensionValue2015,
           }),
-          id: buildInstanceIdentifiersFromIdentifier(observation.identifier),
+          id: {
+            [_rootAttribute]: placeholderOrgOid,
+            [_extensionAttribute]: observation.id ?? observation.identifier?.[0]?.value ?? "",
+          },
           code: buildCodeCvFromCodeableConcept(observation.code),
           statusCode: withoutNullFlavorObject(observation.status, _codeAttribute),
           effectiveTime: {
             low: withoutNullFlavorObject(effectiveTime, _valueAttribute),
             high: withoutNullFlavorObject(effectiveTime, _valueAttribute),
           },
-          value: buildValueST(observation.valueString),
+          value: buildValueSt(observation.valueString),
         },
       },
     };
   });
 }
+
+export function createTableRowsFromObservation(
```

Comment:
so this function, as opposed to createTableRowsFromMedicationStatements or createTableRowFromCondition, actually creates rowS. Is the reason the createTableRowsAndEntries function is structured the way it is, expecting arrays returned in the callbacks?

Reply:
Yes.. I guess this needs to be polished a bit more

Reply:
Fixed the callbacks and properly mapped the types here